### DISCUSSION TITLE
Proposal: Moving Layer `filter` from props into style

### DISCUSSION THREAD

**Discussion by MichaelDanielTom** [rnmapbox/maps#3253] [General]
In `SymbolLayer`, `ModelLayer`, etc. there exists a [filter property](https://github.com/rnmapbox/maps/blob/main/docs/SymbolLayer.md#filter) that takes an expression. In v11, it is required that you update `filter` on `Layer.swift` from MapboxMaps within an `updateLayer` closure, [As seen in this example](https://docs.mapbox.com/ios/maps/examples/pitch-and-distance/) otherwise it does nothing. Currently, setting the `filter` prop on any layer in rnmapbox doesn't do anything because there's no code that sets it in `RNMBXLayer.swift` when RNMBX_11 is true, but even if we do set it here after parsing the json and converting to an expression, it still doesn't update.

We already have the infra for diffing changes and doing everything in the `updateLayer` closure when changing the style, so why not add it there? From an initial test it does seem to work well in there, and we don't have to have an extra updateLayer closure in `RNMBXLayer.swift`, which could possibly lead to two updateLayer's per render.

Additionally, the base `Layer.swift` doesn't have filter in it, and same with the Android equivalent, so it's not something that applies to every layer like `minZoom`, `maxZoom`, `slot`, etc. have. 

What do you all think? @mfazekas

**Comment by mfazekas** [comment]
@MichaelDanielTom thanks much for looking into it.

I'd prefer opening a bug with a component to reproduce and do the discussion there on how to fix this.

So right now `layout` and `paint` properties of a Layer are in style. See the style spec.js:
https://github.com/rnmapbox/maps/blob/ed18ecd2d84d174662895cd73c4c86bbaf4043a5/style-spec/v8.json#L1202-L1222

https://docs.mapbox.com/help/glossary/layout-paint-property/

For me moving filter to style is an API/interface question. 
I hope we can fix the issue without changing the API/interface.

**Reply by MichaelDanielTom** [reply]
Makes sense! And yea, probably wouldn't want to change the rnmapbox api unnecessarily.
